CONTRIBUTING.md

```diff
@@ -12,19 +12,282 @@
+   additional commits over ammending and force-pushing since it can be difficult
+   to follow code reviews when the commit history changes.
+
+   Commits will be squashed when they're merged.
 
 # Formatting
 
 All code must be formatted with `gofmt` (with the latest Go version) and pass
 `go vet`.
 
+# Style
+
+Please read and follow https://github.com/golang/go/wiki/CodeReviewComments for
+all Go code in this repo.
+
+The following style guidelines are specific to writing Go samples.
+
+## One file per sample
+
+Each sample should be in its own file so the imports used by the sample can
+be included in the region tag.
+
+## Sample file name and directory
+
+The top level directory should be the product the sample is for (e.g.
+`functions` or `dlp`).
+
+Sub-directories can be used to keep different groups of samples for the product
+separate.
+
+Files should be named after the sample in them (e.g. `hello.go`). No need to
+include the product name or "sample" in the filename.
+
+If there are many samples to write in the same directory, use filename prefixes
+to group the files acting on similar types (for example, when writing
+create/update/delete type samples).
+
+## Include imports in region tags
+
+The sample region (e.g. `[START foo]` and `[END foo`]) should include the import
+block.
+
+```go
+// Package hello contains Hello samples.
+package hello
+
+// [START hello]
+import "fmt"
+
+func hello(w io.Writer) {
+	fmt.Fprintln(w, "Hello, World")
+}
+
+// [END hello]
+```
+
+For quickstarts, the region should include the package declaration.
+
+For snippets, the region should _not_ include the package declaration.
+
+## Print to an `io.Writer`
+
+Do not print to `stdout` or `stderr`. Pass `w io.Writer` as the first argument
+to the sample function and print to it with `fmt.Fprintf(w, ...)`.
+
+This pattern matches `http.Handler`s, which print to an `http.ResponseWriter`, normally named
+`w`.
+
+```go
+func hello(w io.Writer) {
+	fmt.Fprintln(w, "Hello, World.")
+}
+```
+
+## Google Cloud Project ID
+
+Quickstarts should use an example project ID or add a project ID flag.
+
+If a project ID is needed, snippets should have a `projectID string` argument.
+
+## Declare a `context.Context` as needed
+
+Don't pass a `context.Context` as an argument. New Go developers may not
+understand where the `ctx` comes from.
+
+```diff
+- func hello(ctx context.Context, w io.Writer) { ... }
++ func hello(w io.Writer) {
++ 	ctx := context.Background()
++ 	// ...
++ }
+```
+
+## Function arguments
+
+There should be as few function arguments as possible. An `io.Writer` and
+project ID are the most common. If you need additional arguments (for example,
+the ID of a resource to get or delete), there should be an example value in the
+body of the sample function.
+
+```go
+// delete deletes the resource identified by resourceID.
+func delete(w io.Writer, resourceID string) error {
+	// resource := fmt.Sprintf("/projects/my-project/resources/%s", resourceID)
```

Comment:
i think we call these "name" - do we need to add common identifiers to the style guide, too?